File: Difficulty: Medium/Word Break/word-break.cpp
//{ Driver Code Starts
//Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function template for C++

// s : given string to search
// dictionary : vector of available strings

class Solution
{
public:
    bool solve(int ind,string &s,int len,unordered_map<string,int>&mp){
         if(ind ==len) return true;
         
         for(int i=ind;i<len;i++){
             if(mp.find(s.substr(ind,i-ind+1)) != mp.end()){
                 if(solve(i+1,s,len,mp)) return true;
             }
         }
         return false;
    }
    
    int wordBreak(int n, string s, vector<string> &dictionary) {
        unordered_map<string,int>mp;
        for(auto it:dictionary) mp[it]=1;
        
        return solve(0,s,s.size(),mp);
    }
};

//{ Driver Code Starts.

int main(){
    int t;
    cin>>t;
    while(t--){
        int n;
        cin>>n;
        vector<string> dictionary;
        for(int i=0;i<n;i++){
            string S;
            cin>>S;
            dictionary.push_back(S);
        }
        string s;
        cin>>s;
        Solution ob;
        cout<<ob.wordBreak(n, s, dictionary)<<"\n";
    }
}

// } Driver Code Ends

File: Difficulty: Medium/Word Break/word-break.py
#User function Template for python3

class Solution:
    def solve(self, ind, s, n, word_dict):
        if ind == n:  
            return True
        
        for i in range(ind, n):
            sub_str = s[ind:i + 1]  
            if sub_str in word_dict:  
                if self.solve(i + 1, s, n, word_dict):
                    return True
        
        return False  
        
    def wordBreak(self, n, s, dictionary):
        word_dict = set(dictionary)  # Use a set for O(1) lookups
        if self.solve(0, s, len(s), word_dict):
            return 1
        else:
            return 0



#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    test_case = int(input())

    for _ in range(test_case):
        n = int(input())
        dictionary = [word for word in input().strip().split()]
        s = input().strip()
        ob = Solution()
        res = ob.wordBreak(n, s, dictionary)
        if res:
            print(1)
        else:
            print(0)

# } Driver Code Ends